Docs caching (#7553)
solo-io/gloo: docs/content/guides/traffic_management/listener_configuration/caching/setup.md

---
title: Set up caching
weight: 50
description: Deploy the caching server and start caching responses from upstream services. 
---

Set up the Gloo Edge caching server to cache responses from upstream services for quicker response times.

{{% notice note %}}

> This feature is available only for Gloo Edge Enterprise v1.12.x and later.

Reviewer: maybe this should also be on the about page?

{{% /notice %}}

> When you enable caching during installation, the caching server deployment is automatically created for you and is managed by Gloo Edge. Then, you must configure an HTTP or HTTPS listener to cache responses for its upstream services. When the listener routes a request to an upstream, the response from the upstream is automatically cached for one hour, unless otherwise specified by sending a `cache-control` request header. All subsequent requests receive the cached response.

Reviewer 2: Could we say something like
> When caching is enabled during installation, a deployment is automatically created for you and is managed by Gloo Edge. Then, you must configure an HTTP or HTTPS listener to cache responses for its upstream services. When the listener routes a request to an upstream, the response from the upstream is automatically cached if it contains a `cache-control` repsonse header. All subsequent requests receive the cached response, until the cache entry expires.

## Deploy the caching server

Create a caching server during Gloo Edge Enterprise installation time, and specify any Redis overrides. 

1. [Install Gloo Edge Enterprise version 1.12.x or later by using Helm]({{% versioned_link_path fromRoot="/installation/enterprise/#customizing-your-installation-with-helm" %}}). In your `values.yaml` file, specify the following settings:
   
   {{% notice note %}} 

>    Due to a known issue in Gloo Edge Enterprise 1.12.x, make sure to use the current 1.13.0-rc3 release to try out caching. Run the following command to install Gloo Edge: `helm install -n gloo-system --create-namespace gloo-ee gloo-ee/gloo-ee -f values.yaml --devel --version v1.13.0-rc3 --set-string license_key=<license-key>`

Reviewer: Is this something that will be fixed in 1.12 soon? Otherwise maybe change the notes and docs to just say 1.13 or later

Author: yeah it is supposed to be fixed soonish. ben is working on a fix right now

Reviewer 2: The fix is here: https://github.com/solo-io/solo-projects/pull/4365
We need one more approver and then I will put up a release as soon as this merges

   replicaset.apps/caching-service-5d7f867cdc            1         1         1       76s
   ```

3. Optional: You can also check the debug logs to verify that caching is enabled.
   ```sh
   glooctl debug logs
   ```
   Search the output for `caching` to verify that you have log entries similar to the following:
   ```json
   {"level":"info","ts":"2022-08-02T20:47:30.057Z","caller":"radix/server.go:31","msg":"Starting our basic redis caching service","version":"1.12.0"}
   {"level":"info","ts":"2022-08-02T20:47:30.057Z","caller":"radix/server.go:35","msg":"Created redis pool for caching","version":"1.12.0"}
   ```

<!-- future work
## Configure settings for the caching server

should be able to configure general settings for the server in the future, like the default caching time

https://docs.solo.io/gloo-edge/master/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/enterprise/options/caching/caching.proto.sk/#settings
-->

## Configure caching for a listener

Configure your gateway to cache responses for all upstreams that are served by a listener. Enabling caching for a specific upstream is currently not supported.

> 1. Edit the Gateway CRD where your listener is defined.

Reviewer: I think this is the CR not CRD
```suggestion
1. Edit the Gateway custom resource where your listener is defined.
```

- **Envoy caching service**: The `/valid-for-minute` endpoint is used to show how caching works in Gloo Edge Enterprise with response validation. 

Follow the steps to set up `httpbin` and the Envoy caching service, and to try out caching in Gloo Edge Enterprise. 

1. Deploy and configure `httpbin`. 
   1. Create the `httpbin` namespace and deploy the app. 
      ```shell
      kubectl create ns httpbin
      kubectl -n httpbin apply -f https://raw.githubusercontent.com/solo-io/gloo-mesh-use-cases/main/policy-demo/httpbin.yaml
      ```
   
   2. Verify that the app is up and running. 
      ```shell
      kubectl get pods -n httpbin
      ```
   
      Example output: 
      ```
      httpbin-847f64cc8d-9kz2d   1/1     Running   0          35s
      ```
   
   3. Create an upstream for `httpbin`. 
      ```shell
      kubectl apply -f- <<EOF
      apiVersion: gloo.solo.io/v1
      kind: Upstream
      metadata:
        name: "httpbin"
        namespace: "gloo-system"
      spec:
        static:
          hosts:
          - addr: "httpbin.org"
            port: 80
      EOF
      ```

2. Deploy and configure the Envoy caching service. 

>    1. Create a namespace for the envoy caching service. 

Reviewer: ```suggestion
   1. Create a namespace for the Envoy caching service. 
```

        "url": "http://34.173.214.185/cache/30"
      }
      {{< /highlight >}}
   
   2. Send another request to the same endpoint within the 30s timeframe. In your CLI output, verify that you get back the original response. In addition, check that an `age` response header is returned indicating the age of the cached response, and that the `date` header uses the date and time of the original response. 
      ```shell
      curl -vik "$(glooctl proxy url)/httpbin/cache/30"
      ```
      
      Example output: 
      {{< highlight yaml "hl_lines=2-4" >}}
      ...
      date: Wed, 14 Dec 2022 19:32:13 GMT
      < age: 24
      age: 24

      < 
      {
        "args": {}, 
        "headers": {
        "Accept": "*/*", 
        "Host": "34.173.214.185", 
        "If-Modified-Since": "Wed, 14 Dec 2022 19:03:15 GMT", 
        "User-Agent": "curl/7.77.0", 
        "X-Amzn-Trace-Id": "Root=1-639a24bd-368eb5d92130a8b35144ce4d", 
        "X-Envoy-Expected-Rq-Timeout-Ms": "15000", 
        "X-Envoy-Original-Path": "/httpbin/cache/30"
      }, 
        "origin": "32.200.10.110", 
        "url": "http://34.173.214.185/cache/30"
      }
      {{< /highlight >}}
      

>    3. Wait until the 30 seconds have passed and the cached response becomes stale. Send another request to the same endpoint. Verify that you get back a fresh response and that no `age` header is returned. 

Reviewer: ```suggestion
   3. Wait until the 30 seconds pass and the cached response becomes stale. Send another request to the same endpoint. Verify that you get back a fresh response and that no `age` header is returned. 
```

      < x-envoy-upstream-service-time: 5
      x-envoy-upstream-service-time: 5

      < 
      This response will stay fresh for one minute

      Response generated at: Thu, 15 Dec 2022 15:45:19 GMT
      ```
      
   2. Send another request to the same endpoint within the 1 minute timeframe. Because the response is cached for 1 minute, the original response is returned with an `age` header indicating the number of seconds that have passed since the original response was sent. Make sure that the `date` header and response body include the same information as in the original response. 
      ```shell
      curl -vik "$(glooctl proxy url)/service/1/valid-for-minute"
      ```
      
      Example output: 
      ```
      ...
      < cache-control: max-age=60
      cache-control: max-age=60
      < content-length: 99
      content-length: 99
      < date: Thu, 15 Dec 2022 15:45:19 GMT
      date: Thu, 15 Dec 2022 15:45:19 GMT
      < content-type: text/html; charset=utf-8
      content-type: text/html; charset=utf-8
      < age: 5
      age: 5

      < 
      This response will stay fresh for one minute

      Response generated at: Thu, 15 Dec 2022 15:45:19 GMT
      ```
      

>    3. After the 1 minute has passed and the cached response becomes stale, send another request to the same endpoint. The Envoy caching app is configured to automatically add the `If-Modified-Since` header to each request to trigger the response validation process. In addition, the app is configured to always return a 304 Not Modified HTTP response code to indicate that the response has not changed. When the 304 HTTP response code is received by the Gloo Edge caching server, the caching server fetches the original response from Redis, and sends it back to the client. 

Reviewer: ```suggestion
   3. After the 1 minute passes and the cached response becomes stale, send another request to the same endpoint. The Envoy caching app is configured to automatically add the `If-Modified-Since` header to each request to trigger the response validation process. In addition, the app is configured to always return a `304 Not Modified` HTTP response code to indicate that the response has not changed. When the `304` HTTP response code is received by the Gloo Edge caching server, the caching server fetches the original response from Redis, and sends it back to the client. 
```

>       You can verify that the response validation succeeded when the `date` response header is updated with the time and date of your new request, the `age` response header is removed, and the response body contains the same information as the in the original response. 

Reviewer: ```suggestion
      You can verify that the response validation succeeded when the `date` response header is updated with the time and date of your new request, the `age` response header is removed, and the response body contains the same information as in the original response. 
```

      ```shell
      curl -vik "$(glooctl proxy url)/service/1/valid-for-minute"
      ```
      
      Example output: 
      ```
      ...
      < date: Thu, 15 Dec 2022 15:53:55 GMT
      date: Thu, 15 Dec 2022 15:53:55 GMT
      < server: envoy
      server: envoy
      < x-envoy-upstream-service-time: 5
      x-envoy-upstream-service-time: 5
      < content-length: 99
      content-length: 99
      < content-type: text/html; charset=utf-8
      content-type: text/html; charset=utf-8

      < 
      This response will stay fresh for one minute

      Response generated at: Thu, 15 Dec 2022 15:45:19 GMT
      ```
            
{{% notice note %}}

> Because the Envoy caching app is configured to always return a 304 HTTP response code, you continue to see the cached response no matter how many requests you send to the app. To reset the app and force the app to return a fresh response, you must restart the pod. 

Reviewer: ```suggestion
Because the Envoy caching app is configured to always return a `304` HTTP response code, you continue to see the cached response no matter how many requests you send to the app. To reset the app and force the app to return a fresh response, you must restart the pod. 
```
